src/hashes.rs

```diff
@@ -1,473 +1,313 @@
-        Self::CODE
-    }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
-    }
-}
-impl Blake2b256 {
-    /// The code of Blake2b-256 hasher, 0xb220.
-    pub const CODE: Code = Code::Blake2b256;
-    /// Hash some input and return the blake2b-256 digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        let digest = Blake2b::new()
-            .hash_length(32)
-            .to_state()
-            .update(&data)
-            .finalize();
-        wrap(Self::CODE, &digest.as_bytes())
-    }
-}
-
-/// The Blake2b-512 hasher.
-#[derive(Clone, Debug)]
-pub struct Blake2b512;
-impl MultihashDigest for Blake2b512 {
-    fn code(&self) -> Code {
-        Self::CODE
-    }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
-    }
-}
-impl Blake2b512 {
-    /// The code of Blake2b-512 hasher, 0xb240.
-    pub const CODE: Code = Code::Blake2b512;
-    /// Hash some input and return the blake2b-512 digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        let digest = Blake2b::new()
-            .hash_length(64)
-            .to_state()
-            .update(&data)
-            .finalize();
-        wrap(Self::CODE, &digest.as_bytes())
-    }
-}
-
-/// The Blake2s-128 hasher.
-#[derive(Clone, Debug)]
-pub struct Blake2s128;
-impl MultihashDigest for Blake2s128 {
-    fn code(&self) -> Code {
-        Self::CODE
-    }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
-    }
-}
-impl Blake2s128 {
-    /// The code of Blake2s-128 hasher, 0xb250.
-    pub const CODE: Code = Code::Blake2s128;
-    /// Hash some input and return the blake2s-128 digest.
+impl Identity {
+    /// The code of the Identity hasher, 0x00.
+    pub const CODE: Code = Code::Identity;
+    /// Hash some input and return the raw binary digest.
     pub fn digest(data: &[u8]) -> Multihash {
-        let digest = Blake2s::new()
-            .hash_length(16)
-            .to_state()
-            .update(&data)
-            .finalize();
-        wrap(Self::CODE, &digest.as_bytes())
+        if (data.len() as u64) >= u64::from(std::u32::MAX) {
+            panic!("Input data for identity hash is too large, it needs to be less than 2^32.")
+        }
+        wrap(Self::CODE, data)
     }
 }
 
-/// The Blake2s-256 hasher.
-#[derive(Clone, Debug)]
-pub struct Blake2s256;
-impl MultihashDigest for Blake2s256 {
-    fn code(&self) -> Code {
-        Self::CODE
-    }
-    fn digest(&self, data: &[u8]) -> Multihash {
-        Self::digest(data)
-    }
-}
-impl Blake2s256 {
-    /// The code of Blake2s-256 hasher, 0xb260.
-    pub const CODE: Code = Code::Blake2s256;
-    /// Hash some input and return the blake2s-256 digest.
-    pub fn digest(data: &[u8]) -> Multihash {
-        let digest = Blake2s::new()
-            .hash_length(32)
-            .to_state()
-            .update(&data)
-            .finalize();
-        wrap(Self::CODE, &digest.as_bytes())
-    }
+derive_digest! {
```

Comment:
I'm not sure about it myself, but would it make sense so split `derive_digest` into `derive_digest_sha` and `derive_digest_blake`? They don't seem to share code and they use different parameters. This might make the code easier to understand. What do you think?

Reply:
It'll ultimately look and behave the same - the only change would be swapping `@sha` with `sha` in the name of the macro.

Reply:
If you care only about one of the macros while reading the source, it would be less cognitive load. Anyway, it's a matter of taste and there isn't a right or wrong. I'm happy to keep it as it currently is.

I'm using code review also as a chance to learn coding taste/decisions from others, hence I thought I also share mine :)

Reply:
I appreciate the flexibility :)